stl/inc/span

```diff
@@ -0,0 +1,665 @@
+        _STL_VERIFY(_Off < size(), "span index out of range");
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+        return _Mydata[_Off];
+    }
+
+    _NODISCARD constexpr reference front() const noexcept // strengthened
+    {
+        if constexpr (_Extent != dynamic_extent) {
+            static_assert(_Extent > 0, "front of empty span");
+        }
+#if _CONTAINER_DEBUG_LEVEL > 0
+        else {
+            _STL_VERIFY(size() > 0, "front of empty span");
+        }
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+        return _Mydata[0];
+    }
+
+    _NODISCARD constexpr reference back() const noexcept // strengthened
+    {
+        if constexpr (_Extent != dynamic_extent) {
+            static_assert(_Extent > 0, "back of empty span");
+        }
+#if _CONTAINER_DEBUG_LEVEL > 0
+        else {
+            _STL_VERIFY(size() > 0, "back of empty span");
+        }
+#endif // _CONTAINER_DEBUG_LEVEL > 0
+        return _Mydata[size() - 1];
+    }
+
+    _NODISCARD constexpr pointer data() const noexcept {
+        return _Mydata;
+    }
+
+    // [span.iterators] Iterator support
+    _NODISCARD constexpr iterator begin() noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        return iterator(_Mydata, _Mydata, _STD next(_Mydata, size()));
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+        return iterator(_Mydata);
+#endif // _ITERATOR_DEBUG_LEVEL
+    }
+
+    _NODISCARD constexpr iterator end() noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        return iterator(_Mydata + size(), _Mydata, _Mydata + size());
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+        return iterator(_Mydata + size());
+#endif // _ITERATOR_DEBUG_LEVEL
+    }
+
+    _NODISCARD constexpr const_iterator begin() const noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        return const_iterator(_Mydata, _Mydata, _Mydata + size());
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+        return const_iterator(_Mydata);
+#endif // _ITERATOR_DEBUG_LEVEL
+    }
+
+    _NODISCARD constexpr const_iterator end() const noexcept {
+#if _ITERATOR_DEBUG_LEVEL >= 1
+        return const_iterator(_Mydata + size(), _Mydata, _Mydata + size());
+#else // ^^^ _ITERATOR_DEBUG_LEVEL >= 1 ^^^ // vvv _ITERATOR_DEBUG_LEVEL == 0 vvv
+        return const_iterator(_Mydata + size());
+#endif // _ITERATOR_DEBUG_LEVEL
+    }
+
+#ifdef __cpp_lib_concepts
+    _NODISCARD friend constexpr const_iterator begin(span _Right) noexcept {
+        // non-member overload to model the exposition-only forwarding-range concept
+        return _Right.begin();
+    }
+    _NODISCARD friend constexpr const_iterator end(span _Right) noexcept {
+        // non-member overload to model the exposition-only forwarding-range concept
+        return _Right.end();
+    }
+#endif // __cpp_lib_concepts
+
+    _NODISCARD constexpr const_iterator cbegin() const noexcept {
+        return begin();
+    }

```

Comment:
Does this compile? My brain-compiler says no - you are converting `iterator` to `const_iterator`, but there is no conversion between `_Span_iterator<_Ty>` and `_Span_iterator<const _Ty>`, which is the true problem. (Also be aware that https://eel.is/c++draft/container.requirements#general-7 is a requirement.)

Many containers currently have their `iterator` derive from their `const_iterator` but this is undesirable; there should be a conversion instead of an inheritance relationship.

Reply:
As far as I can tell there is no conversion as `cbegin()` is a `const` member method so it falls back to the `begin() const` member method wich already returns a `const_iterator`